### PR TITLE
[15.0][ADD] mis_builder_budget_product

### PR DESCRIPTION
**Pull request by Camille0907**
- [x] Depends on https://github.com/OCA/mis-builder-contrib/pull/33 to fix pre-commit 

We want to be able to choose a specific product in a MIS Budget by account item.

### CODE REVIEW

**Review by cyrilmanuel — APPROVED** [review]

**Review by yankinmax — APPROVED** [review]

File: mis_builder_budget_contrib/models/mis_budget_by_account_item.py
Comment (SilvioC2C) [inline mis_builder_budget_contrib/models/mis_budget_by_account_item.py]: Can you please add a comment explaining the expected behavior if `product_id` is not defined on `self` or, viceversa, why we can expect it to be always properly filled?

**Review by SilvioC2C — COMMENTED** [review]
Just a small comment

File: mis_builder_budget_contrib/models/mis_budget_by_account_item.py
Comment (Camille0907) [reply mis_builder_budget_contrib/models/mis_budget_by_account_item.py]: If the `product_id` is not defined, it was detecting overlapping with another empty line (`product_id` not defined <=> `product_id = product.product()` <=> `product_id.id = product.product().id = False`)

But thanks to your remark I saw another issue : it was not counting overlapping for a line with no product (including all products) and a line with one product.

Changed accordingly, so I answer you here instead of in comments

File: mis_builder_budget_contrib/readme/HISTORY.rst
Comment (simahawk) [inline mis_builder_budget_contrib/readme/HISTORY.rst]: if you don't want to maintain this file you can use newsfragments https://github.com/OCA/maintainer-tools/tree/master/template/module#automatic-changelog-generation

File: mis_builder_budget_contrib/readme/HISTORY.rst
Comment (simahawk) [inline mis_builder_budget_contrib/readme/HISTORY.rst]: new module?

File: mis_builder_budget_contrib/models/mis_budget_by_account_item.py
Comment (simahawk) [inline mis_builder_budget_contrib/models/mis_budget_by_account_item.py]: maybe better to use `expressions.AND`?

File: mis_builder_budget_contrib/models/mis_budget_by_account_item.py
Comment (simahawk) [inline mis_builder_budget_contrib/models/mis_budget_by_account_item.py]:
```suggestion
    @api.constrains("product_id")
    def _check_product(self):
        """Check dates if product changes."""
        self._check_dates()
        return
```

Comment (Camille0907) [reply mis_builder_budget_contrib/models/mis_budget_by_account_item.py]: If I do it, would the method be call if th analytic_account_id (for example) is modified ?

File: mis_builder_budget_contrib/readme/HISTORY.rst
Comment (Camille0907) [reply mis_builder_budget_contrib/readme/HISTORY.rst]: https://github.com/OCA/mis-builder/commit/f861cd22faded606bcbff4a67d7bf8b5fff72ee4
This way ?

File: mis_builder_budget_contrib/models/mis_budget_by_account_item.py
Comment (Camille0907) [reply mis_builder_budget_contrib/models/mis_budget_by_account_item.py]:
```
domain = expression.AND(domain, [ "|", ("product_id", "=", False), ("product_id", "=", self.product_id.id)]
```
?
Why is it better ? :)

File: mis_builder_budget_contrib/readme/HISTORY.rst
Comment (simahawk) [reply mis_builder_budget_contrib/readme/HISTORY.rst]: hmm, no, you don't need all that work. OCA quality tools will take care of rendering, you just have to put file for your change in `newsfragments` folder.
NOTE: the changelog is not mandatory, you can skip it.

File: mis_builder_budget_contrib/models/mis_budget_by_account_item.py
Comment (simahawk) [reply mis_builder_budget_contrib/models/mis_budget_by_account_item.py]: Yes. Main reasons:

1. it's more explicit what you want
2. you don't know what's coming from the super call and you might be altering something you don't want

File: mis_builder_budget_contrib/models/mis_budget_by_account_item.py
Comment (simahawk) [reply mis_builder_budget_contrib/models/mis_budget_by_account_item.py]: What's your goal? Trigger `_check_dates` when the product is touched, correct?
Then this way you don't alter the original method at all and you avoid replicating all the constrained fields which might change somewhere else.

**Review by fclementic2c — APPROVED** [review]
Functional test ok

**Review by simahawk — APPROVED** [review]
LG but I'm not sure about  the module name. @sbidoul do you have any preference?

**Comment by yvaucher** [comment]
ping @sbidoul about the module name

**Comment by sbidoul** [comment]
I think it's best to avoid catch-all module names, no? How about `mis_builder_budget_product`?

**Comment by yvaucher** [comment]
Renamed

**Comment by rafaelbn** [comment]
@yvaucher 😄 you  have renamed the module in the PR but the module is https://github.com/OCA/mis-builder-contrib/tree/15.0/mis_builder_budget_contrib with the original one. Do you know why?

**Comment by yvaucher** [comment]
@rafaelbn I have no clue, does OCA-git-bot didn't pull the right version when merging ? I'll rename it manually.

**Comment by sbidoul** [comment]
This is really weird indeed.

**Comment by yvaucher** [comment]
@sbidoul @rafaelbn I pushed force to make sure it's the right commit. Thanks for the heads up.

@sbidoul I didn't wait for the CI to be green on this PR to launch the command of the oca bot after I pushed expecting the oca bot to do the checking anyway. I might have been too hasty and the git branch was not up to date. (I think that Github update first the info on the UI and push to git afterwards but I could be wrong)

**Comment by yvaucher** [comment]
It was detected I pushed on this branch at 4:16
And launched the command and @OCA-git-bot pulled it at 4:17

The sequence is right but 1 min might not be enough for the code to be up to date :confused: 

Next time I'll drink a tea in the mean time ^^

**Comment by sbidoul** [comment]
The bot does regular git commands, so that really should not happen unless there are multiple replicas of the git repo and some replication delay?

**Pull request closed without merging**
